add project interfaces, services, models, DTOs, validators

mohammadsofan opened the pull request:
add project interfaces, services, models, DTOs, validators.

mohammadsof committed:
feat: initial commit - add project interfaces, services, models, DTOs, and validators

MohammadBDiab commented on InventoryApp/Dtos/GetProductsResult.cs:
```diff
@@ -0,0 +1,10 @@
+
+namespace InventoryApp.Dtos
+{
+    internal class GetProductsResult
```
- GetProductsResult is fine, but consider using a suffix like Dto to clearly mark it as a Data Transfer Object, may you can rename it to GetProductsResultDto 
- Returning null for Products might require extra null checks. Often it's better to return an empty list, its best to Initialize Products to an empty list

MohammadBDiab commented on InventoryApp/Dtos/DeleteResult.cs:
```diff
@@ -0,0 +1,8 @@
+{
+        internal class DeleteResult 
+        {
+            public string Message { get; set; } = null!;
```
Your used Null-forgiving Operator (null!)
- Using = null!; silences nullability warnings but can lead to runtime exceptions if not set properly.
- It best to provide a default value or use nullable type:

- Option A (safe default):
public string Message { get; set; } = string.Empty;

- Option B (nullable + null check where used):
public string? Message { get; set; }

MohammadBDiab commented on InventoryApp/Dtos/GetProductResult.cs:
```diff
@@ -0,0 +1,10 @@
+
+namespace InventoryApp.Dtos
+{
+    internal class GetProductResult
```
GetProductResultDto may is better , consider using a suffix like Dto to clearly mark it as a Data Transfer Object

MohammadBDiab commented on InventoryApp/Dtos/GetProductResult.cs:
```diff
@@ -0,0 +1,10 @@
+    internal class GetProductResult
+    {
+        public Product? Product { get; set; }
+        public string? Message {  get; set; }
```
If Message is always expected, consider initializing it with an empty string instead of making it nullable.

MohammadBDiab commented on InventoryApp/Dtos/Result.cs:
```diff
@@ -0,0 +1,11 @@
+
+namespace InventoryApp.Dtos
+{
+    internal class Result
```
**Result** is too broad and may conflict with other common types (e.g., System.Result, third-party libraries, or other custom types).
Use a more descriptive name:
- OperationResult
- ProductOperationResult

MohammadBDiab commented on InventoryApp/Dtos/Result.cs:
```diff
@@ -0,0 +1,11 @@
+{
+    internal class Result
+    {
+        public string Message { get; set; } = null!;
```
Avoid null! unless absolutely necessary. Instead, initialize with a safe default:

MohammadBDiab commented on InventoryApp/Interfaces/IFileService.cs:
```diff
@@ -0,0 +1,11 @@
+
+namespace InventoryApp.Interfaces
+{
+    internal interface IFileService
```
Currently tied to Product. Consider making it generic for reusability across different data types - IFileService<T>

MohammadBDiab commented on InventoryApp/Interfaces/IFileService.cs:
```diff
@@ -0,0 +1,11 @@
+{
+    internal interface IFileService
+    {
+        void WriteToFile(IList<Product> products);
```
If you expect to work with multiple files, it’s better to allow a file path input instead of hardcoding internally.

MohammadBDiab commented on InventoryApp/Interfaces/IProductService.cs:
```diff
@@ -0,0 +1,17 @@
+
+namespace InventoryApp.Interfaces
+{
+    internal interface IProductService
```
Although not strictly necessary, method names like GetAllProducts, GetProductByFilter, or UpdateProduct can be clearer than GetProducts, GetProduct, and Update.

MohammadBDiab commented on InventoryApp/Models/Product.cs:
```diff
@@ -0,0 +1,35 @@
+namespace InventoryApp.Models
+{
+    internal class Product
+    {
```
You're using = null!; on Name and Description, which is okay, but it’s safer to initialize them or make them nullable.
public string Name { get; set; } = string.Empty;
public string Description { get; set; } = string.Empty;
OR 
public string? Name { get; set; }
public string? Description { get; set; }

MohammadBDiab commented on InventoryApp/Models/Product.cs:
```diff
@@ -0,0 +1,35 @@
+        public string Description { get; set; } = null!;
+        public double Price { get; set; }
+        public double Discount { get; set; }
+        public double FinalPrice => Price - Price * Discount; 
```
If you plan to show prices to users or in reports, it may be useful to round to 2 decimal places to avoid floating-point artifacts.
public double FinalPrice => Math.Round(Price - Price * Discount, 2);

MohammadBDiab commented on InventoryApp/Program.cs:
```diff
@@ -0,0 +1,195 @@
+                    """);
+        }
+        public static Product ReadProductInfo()
+        {
```
- this function is too long should be split to multi methods 
- Consider moving the menu logic to a class like InventoryAppMenu for cleaner separation of concerns.

You use Console.ReadLine()! in ReadProductInfo, which is okay but could throw if ReadLine() returns null (e.g., redirected input).

Recommended safer version:
product.Name = Console.ReadLine() ?? string.Empty;

MohammadBDiab commented on InventoryApp/Program.cs:
```diff
@@ -0,0 +1,195 @@
+            }
+            switch(option)
+            {
+                case 1:
```
please use Enum instead of number case 1 -> case Opertion.CreateProduct

MohammadBDiab commented on InventoryApp/Program.cs:
```diff
@@ -0,0 +1,195 @@
+                    var result = productService.Delete(id);
+                    Console.WriteLine(result.Message);
+                    goto displayOptionsList;
+                }
```
Before deleting, prompt the user to confirm:
Console.Write("Are you sure you want to delete this product? (y/n): ");
var confirmation = Console.ReadLine();
if (confirmation?.ToLower() != "y") {
    Console.WriteLine("Deletion cancelled.");
    break;
}

MohammadBDiab commented on InventoryApp/Services/FileService.cs:
```diff
@@ -0,0 +1,64 @@
+        public FileService(string path)
+        {
+            FilePath = path;
+            if (!Directory.Exists(Path.GetDirectoryName(FilePath)))
```
Path.GetDirectoryName(FilePath) may return null ?

MohammadBDiab commented on InventoryApp/Services/FileService.cs:
```diff
@@ -0,0 +1,64 @@
+        }
+        public void WriteToFile(IList<Product> products)
+        {
+            try
```
Use of JsonSerializerOptions for Formatting (Optional)
If you want readable JSON (e.g., for debugging), format it:
var json = JsonSerializer.Serialize(products, new JsonSerializerOptions { WriteIndented = true });

MohammadBDiab commented on InventoryApp/Services/ProductService.cs:
```diff
@@ -0,0 +1,234 @@
+        {
+            try
+            {
+                var validatorResult = validator.Validate(product);
```
Rename validatorResult to validationResult for clarity.

MohammadBDiab commented on InventoryApp/Services/ProductService.cs:
```diff
@@ -0,0 +1,234 @@
+            try
+            {
+                var products = fileService.ReadFromFile();
+                var oldProduct = products.FirstOrDefault(p => p.Id == id);
```
In Update, consider renaming oldProduct to existingProduct.

MohammadBDiab commented on InventoryApp/Services/ProductService.cs:
```diff
@@ -0,0 +1,234 @@
+            try
+            {
+                var products = GetProducts(filter).Products;
+                Product? product = null;
```
var product = products?.FirstOrDefault();

MohammadBDiab commented on InventoryApp/Validators/ProductValidator.cs:
```diff
@@ -0,0 +1,55 @@
+    public class ProductValidatorResult
+    {
+        public bool IsValid { get; set; }
+        public IList<Error> Errors { get; set; } = null!;
```
Using = null!; is risky — always initialize collections to avoid NullReferenceException.

public IList<Error> Errors { get; set; } = new List<Error>();

MohammadBDiab commented on InventoryApp/Validators/ProductValidator.cs:
```diff
@@ -0,0 +1,55 @@
+                    Message = "Discount must be between 0.00 and 1.00."
+                });
+                isValid = false;
+            }
```
You’re setting isValid = false every time you add an error, but that can be inferred at the end.
return new ProductValidatorResult
{
    IsValid = !errors.Any(),
    Errors = errors
};

MohammadBDiab commented on InventoryApp/Validators/ProductValidator.cs:
```diff
@@ -0,0 +1,55 @@
+        public string Field { get; set; } = null!;
+        public string Message { get; set; } = null!;
+    }
+    internal class ProductValidator
```
Consider adding validation for:
- Max name/description length
- Price and quantity non-negativity

mohammadsof committed:
feat: Add Serilog and apply logging in services and main method

mohammadsof committed:
refactor: clean up and rename variables for clarity

mohammadsof committed:
refactor: replace null! with nullable types or default values

mohammadsof committed:
refactor: improve null safety and code readability, add JSON formatting

mohammadsof committed:
refactor: replace numeric switch with enum and improve validation logic

mohammadsof committed:
refactor: convert FileService and IFileService to generic types

mohammadsof committed:
refactor(file-service): make file path configurable per operation instead of fixed

- Removed fixed FilePath from FileService<T>
- Added filePath as a parameter to WriteToFile and ReadFromFile
- Improved logging with more descriptive messages and error handling
- Updated ProductService to pass the path on each file operation

mohammadsof committed:
refactor(program): separate input, output and menu logic into helper classes

- Extract user input methods into ConsoleInputHelper
- Move product printing logic to ConsoleOutputHelper
- Move menu display to InventoryAppMenu
- Update Program.cs to use these helpers for cleaner and maintainable code

mohammadsof committed:
feat(input): add confirmation prompt before product deletion

- Added ConfirmDeletion method to ConsoleInputHelper
- Prompt user to confirm deletion before proceeding
- Prevent accidental product deletion

mohammadsof committed:
feat(logging): restructured Serilog using appsettings.json with support for text, JSON, and Seq sinks

- Moved Serilog configuration to appsettings.json
- Added rolling file logging in both .txt and .json formats
- Integrated logging with Seq at http://localhost:5341/
- Removed old hardcoded logger setup

MohammadDiab commented:
Approved
